src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2258,177 @@ int64_t BlueFS::_read_random(
   return ret;
 }
 
+void BlueFS::_wal_update_size(FileRef file, uint64_t flush_offset, uint64_t increment) {
+  dout(20) 
+      << fmt::format(
+        "{} updating WAL file {} for range {:#x}~{:#x} limit is {:#x}", 
+        __func__, file->fnode.ino, flush_offset, increment, file->fnode.wal_limit) 
+      << dendl;
+  ceph_assert(file->is_wal_read_loaded);
+  ceph_assert((flush_offset == 0 && file->wal_flushes.empty()) || (flush_offset == file->wal_flushes.back().end_offset()));
+
+  FileReader *h = new FileReader(file, cct->_conf->bluefs_max_prefetch, false, true);
+
+  uint64_t flush_end = flush_offset + increment;
+  while (flush_offset < file->fnode.wal_limit) {
+    // read first part of wal flush
+    bufferlist bl;
+    uint64_t read_result = (uint64_t)_read(h, flush_offset, sizeof(File::WALFlush::WALLength), &bl, nullptr);
+    if (read_result < sizeof(File::WALFlush::WALLength)) {
+      dout(20) << fmt::format("{} cannot read flush length, most likely we are out of bounds. flush_offset={:#X}", __func__, flush_offset) << dendl;
+      break;
+    }
+    File::WALFlush::WALLength flush_length_le(0);
+    flush_length_le = *((File::WALFlush::WALLength*)bl.c_str());
+    uint64_t flush_length = flush_length_le;
+    dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+    // read marker
+    bl.clear();
+    uint64_t marker_offset = flush_offset+sizeof(File::WALFlush::WALLength)+flush_length;
+    read_result = _read(h, marker_offset, sizeof(File::WALFlush::WALMarker), &bl, nullptr);
+    File::WALFlush::WALMarker marker_le(0);
+    if (read_result < sizeof(File::WALFlush::WALMarker)) {
+      dout(20) << fmt::format("{} cannot read marker, most likely we are out of bounds. flush_offset={:#X}, marker_offset={:#X}", __func__, flush_offset, marker_offset) << dendl;
+      break;
+    }
+    marker_le = *((File::WALFlush::WALMarker*)bl.c_str());
+    uint64_t marker = marker_le;
+    if (marker != file->fnode.ino) {
```

Comment:
WALv2 is relaying on data it did not write to extract information.
With any method like this, we need sufficiently low probability of false positive detection.
False positive here means that we:
- read data we did not write
- decided that the data is legit

I will try to estimate how reliable is "ino + uuid" as a marker check.
1) UUID
How many uuid sequences are on disk?
Lets assume uuid is fully random, meaning that probability of random bits forming uuid is 0.5^(8*8). With that, we do not expect even one uuid to appear on the device.
But uuid is written once per `bluefs_transaction_t`. We will have many uuids from this source.
Lets assume bluefs log and rocksdb sst files are the only contenders for disk space.
In the system running long time, all disk has been used. The balance between (conf.bytes_written_wal + conf.bytes_written_sst) and  (conf.logged_bytes) determines amount of uuid on device. So it will be many, like many thousands. (numbers needed!)
2) INO
How many ino values might be on the disk?
As ino are small integer values, we should expect plenty of other, unrelated, small 32 bit values. I think its a lot.

In `bluefs_transaction_t` right before uuid is `len`, and right after it is `seq`. Both `len` and `seq` we expect to be small integers.
Ino + uuid are a highly unreliable pair for unique marker.

*Proposal*
Lets use hash(ino + uuid) as marker.
If our hash function is good, it will be basically random. There will be no risk of contamination from previous deployments, and we will have no negative side effect from `ino` being small value.
We can tune probability of false possitive simply by taking longer hash function.
For 32bit hash, we expect to have 1 pattern per 4GB of disk.
For 64bit hash, there should be none, except those written specifically to the WALv2.

Reply:
@ifed01 I've implemented a version a hashed generated marker based on Adam's comments. let me know what you think :)